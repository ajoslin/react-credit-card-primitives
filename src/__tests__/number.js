const React = require('react')
const {mount} = require('enzyme')
const Number = require('../number')

function setup ({render = () => <div />, ...props} = {}) {
  let renderArg
  const renderSpy = jest.fn(arg => {
    renderArg = arg
    return render(arg)
  })
  const wrapper = mount(<Number {...props} render={renderSpy} />)
  return {renderSpy, wrapper, ...renderArg}
}

test('value formatting', () => {
  const {wrapper, setValue, getInputProps} = setup()

  setValue('424242')
  expect(getInputProps().value).toEqual('4242 42')

  setValue('4242424242424242')
  expect(getInputProps().value).toEqual('4242 4242 4242 4242')
})

test('validation on all card types', () => {
  const {wrapper, valid, setValue, getInputProps} = setup()

  expect(getInputProps()).toMatchSnapshot('basic no input')
  expect(valid).toBe(false)
  setValue('378282246310005')
  expect(getInputProps()).toMatchSnapshot('basic amex valid')
  expect(wrapper.instance().isValid()).toBe(true)
  setValue('37828224631000')
  expect(getInputProps()).toMatchSnapshot('basic amex invalid')
  expect(wrapper.instance().isValid()).toBe(false)
})

test('validation on given card types', () => {
  const {wrapper, valid, setValue} = setup({
    cardTypes: ['Visa']
  })

  expect(valid).toBe(false)
  setValue('378282246310005')
  expect(wrapper.instance().isValid()).toBe(false)
  expect(wrapper.instance().getStateAndHelpers()).toMatchSnapshot()

  // Make it Visa and it will be valid.
  setValue('4242424242424242')
  expect(wrapper.instance().isValid()).toBe(true)
  expect(wrapper.instance().getStateAndHelpers()).toMatchSnapshot()
})

test('masked', () => {
  const {setValue, getInputProps} = setup({
    masked: true
  })

  // Invalid card (missing one digit)
  setValue('424242424242424')
  expect(getInputProps().value).toEqual('4242 4242 4242 424')

  // Valid AMEX
  setValue('378282246310005')
  expect(getInputProps().value).toEqual('•••• •••••• 10005')
})

test('onChange uncontrolled', () => {
  let changeData
  const {setValue, wrapper} = setup({
    onChange: (data) => {
      changeData = data
    }
  })
  expect(changeData).toBe(undefined)
  setValue('42424')
  expect(changeData.value).toBe('42424')
  expect(changeData.valid).toBe(false)
  expect(changeData.type).toBe('Visa')

  expect(wrapper.state('value')).toEqual('42424')
})

test('onChange w/ controlled input', () => {
  let changeData
  const {wrapper, setValue} = setup({
    value: '',
    onChange: (data) => {
      changeData = data
    }
  })
  expect(changeData).toBe(undefined)
  setValue('42424')
  expect(changeData.value).toBe('42424')
  expect(changeData.valid).toBe(false)
  expect(changeData.type).toBe('Visa')

  // No setState because it's a controlled input
  expect(wrapper.state('value')).toEqual('')
})

test('defaultValue', () => {
  const {wrapper, setValue} = setup({
    defaultValue: '4242'
  })
  expect(wrapper.state('value')).toBe('4242')
  setValue('')
  expect(wrapper.state('value')).toBe('')
})
