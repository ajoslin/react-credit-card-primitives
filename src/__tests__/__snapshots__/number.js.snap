// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic amex invalid 1`] = `
Object {
  "aria-invalid": "true",
  "autoComplete": "cardnumber",
  "maxLength": undefined,
  "name": "cc-number",
  "onBlur": [Function],
  "onChange": [Function],
  "onFocus": [Function],
  "placeholder": "Card number",
  "type": "text",
  "value": "3782 822463 1000",
}
`;

exports[`basic amex valid 1`] = `
Object {
  "aria-invalid": "false",
  "autoComplete": "cardnumber",
  "maxLength": undefined,
  "name": "cc-number",
  "onBlur": [Function],
  "onChange": [Function],
  "onFocus": [Function],
  "placeholder": "Card number",
  "type": "text",
  "value": "3782 822463 10005",
}
`;

exports[`basic no input 1`] = `
Object {
  "aria-invalid": "",
  "autoComplete": "cardnumber",
  "maxLength": undefined,
  "name": "cc-number",
  "onBlur": [Function],
  "onChange": [Function],
  "onFocus": [Function],
  "placeholder": "Card number",
  "type": "text",
  "value": "",
}
`;

exports[`validation on given card types 1`] = `
Object {
  "getInputProps": [Function],
  "setValue": [Function],
  "type": "American Express",
  "valid": false,
  "value": "378282246310005",
}
`;

exports[`validation on given card types 2`] = `
Object {
  "getInputProps": [Function],
  "setValue": [Function],
  "type": "Visa",
  "valid": true,
  "value": "4242424242424242",
}
`;
